dbms/src/Common/FiberRWLock.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/fiber/all.hpp>
+
+namespace DB
+{
+/// FiberRWLock has the same interface with `std::shared_mutex` and is intended to
+/// work with `std::shared_lock` or `std::unique_lock`.
+/// FiberRWLock could work well both on OS thread and boost::fiber, though the
+/// latter performs better.
+class FiberRWLock
+{
+public:
+    FiberRWLock() = default;
+    FiberRWLock(const FiberRWLock &) = delete;
+    ~FiberRWLock() = default;
+
+    FiberRWLock & operator=(const FiberRWLock &) = delete;
+
+    void lock()
+    {
+        std::unique_lock lock(mu);
+        ++writer_wait_cnt;
+        writer_cv.wait(lock, [&] { return writer_cnt == 0 && reader_cnt == 0; });
+        --writer_wait_cnt;
+        writer_cnt = 1;
+    }
+
+    bool try_lock()
+    {
+        std::unique_lock lock(mu);
+        if (writer_cnt == 0 && reader_cnt == 0)
+        {
+            writer_cnt = 1;
+            return true;
+        }
+        return false;
+    }
+
+    void unlock()
+    {
+        std::unique_lock lock(mu);
+        writer_cnt = 0;
+        if (writer_wait_cnt != 0)
+            writer_cv.notify_one();
+        else
+            reader_cv.notify_all();
+    }
+
+    void lock_shared()
+    {
+        std::unique_lock lock(mu);
+        reader_cv.wait(lock, [&] { return writer_cnt == 0 && writer_wait_cnt == 0; });
+        ++reader_cnt;
+    }
+
+    bool try_lock_shared()
+    {
+        std::unique_lock lock(mu);
+        if (writer_cnt == 0 && writer_wait_cnt == 0)
+        {
+            ++reader_cnt;
+            return true;
+        }
+        return false;
+    }
+
+    void unlock_shared()
+    {
+        std::unique_lock lock(mu);
+        if (--reader_cnt == 0 && writer_wait_cnt != 0)
+            writer_cv.notify_one();
+    }
+
+private:
+    boost::fibers::mutex mu;
+    boost::fibers::condition_variable reader_cv;
+    boost::fibers::condition_variable writer_cv;
+    int reader_cnt = 0;
```

Comment:
In class __shared_mutex_base, it only use one unsigned _state_ to track the reader or writer's state,  maybe it is more consice.  ref: https://zhuanlan.zhihu.com/p/463286822

Reply:
It combines two logical states into one field and messes them up. Since the memory usage isn't an issue, I'd like to use two states for better readability.